### PR TITLE
Refuse to overwrite any directory containing the home directory or the profile directory

### PR DESCRIPTION
**Pull request by personalizedrefrigerator**
# Summary

Prevents the plugin from overwriting `os.homedir()` or the Joplin profile directory.

## Rationale

https://github.com/laurent22/joplin/issues/9857 suggests backing up to a subdirectory of the home directory by default. One way to do this is to set Simple Backup's `path` setting to the HOME directory.

Without additional checks, this could be dangerous if a user unchecks the "create subfolder" option.

## Alternatives

- Don't overwrite any directory that contains the profile directory

# Testing

This pull request has a related automated test. However, this only tests one part of this pull request.

To test the pull request manually,
1. Create a new user account
2. Build the plugin from this branch and load it as a development plugin
3. Start Joplin
4. Set the backup path to the home directory
5. Create a backup
6. Disable subfolder creation
7. Try to create a backup
8. Set the backup path to the `.config` directory (e.g. `/home/user/.config/`)
9. Try to create a backup
10. Set the backup path to the profile directory
11. Try to create a backup

This has been tested successfully with commit 2f8ccdab529566988b1a840b0ae98c60bc79133c on OpenSUSE 20240202 with Joplin 2.14.12.

### CODE REVIEW

File: __test__/backup.test.ts [hunk 1]
```diff
@@ -168,7 +168,7 @@ describe("Backup", function () {
     });
 
     it(`relative paths`, async () => {
-      const backupPath = "../";
+      const backupPath = "../foo";
```
Comment (personalizedrefrigerator) [inline __test__/backup.test.ts:171]: Previously, the `backupPath` was a parent directory of the test profile directory. As such, the relative path has been changed.

**Comment by JackGruber** [comment]
Just so I understand it correctly, the PR is supposed to prevent backups from being made without a subfolder in `.config`.

**Comment by personalizedrefrigerator** [comment]
> Just so I understand it correctly, the PR is supposed to prevent backups from being made without a subfolder in` .config`.

The goal of this pull request is to prevent the plugin from deleting Joplin's profile directory or the user's home directory.

## Example 1

Suppose that
1. A user disables "create subfolder"
2. Some time later, the user sets the backup location to `/home/usernamehere/` (the home directory)

This change should prevent the plugin from overwriting `/home/usernamehere/`.

## Example 2

Suppose that
1. A user installs Joplin Portable on a USB drive
    - This would put the `JoplinProfile` directory somewhere on the USB drive
2. The user sets the backup directory to the root of the USB drive (e.g. `D:\`) and disables "create subfolder"
    - This might happen if, for example, the user selects the wrong backup location
3. The user tries to create a backup

Because the root directory of the drive contains the profile directory, this change should prevent the contents of the USB drive from being overwritten.

File: src/helper.ts [hunk 2]
```diff
@@ -65,4 +66,28 @@ export namespace helper {
...
+
+  // Doesn't resolve simlinks
+  // See https://stackoverflow.com/questions/44892672/how-to-check-if-two-paths-are-the-same-in-npm
+  // for possible alternative implementations.
+  export function isSubdirectoryOrEqual(
+    parent: string,
+    possibleChild: string,
+
+    // Testing only
+    pathModule: typeof path = path
```
[the remark this answers is not in the document]
Comment (personalizedrefrigerator) [reply src/helper.ts:78]: I think the difficulties here are that
1. `path.sep` is a readonly-nonfunction property, so we can't `jest.spyOn(path, 'resolve')`
2. `path` is imported at the top of `src/helper.ts`, so `jest.doMock` will only affect the import at the top of the file.
3. `jest.spyOn(path, 'resolve').mockImplementation((...args) => path.posix.resolve(...args))` causes infinite recursion on POSIX systems.

[the remark this answers is not in the document]
Comment (personalizedrefrigerator) [reply src/helper.ts:78]: The third issue above seems to be the most significant.

Even this alternative implementation causes a stack overflow:

https://github.com/personalizedrefrigerator/joplin-plugin-backup-fork/blob/a6f2b93bc7fa7e8f994c1b28d077a891bd61c1aa/__test__/help.test.ts#L196

File: src/Backup.ts [hunk 3]
```diff
@@ -288,11 +289,16 @@ class Backup {
...
-        i18n.__("msg.error.backupPathJoplinDir", path.normalize(profileDir))
-      );
+      await this.showError(i18n.__(errorId, invalidBackupPath));
+    };
+
+    if (helper.isSubdirectoryOrEqual(this.backupBasePath, os.homedir())) {
+      await handleInvalidPath("msg.error.backupPathContainsHomeDir");
+    } else if (helper.isSubdirectoryOrEqual(this.backupBasePath, profileDir)) {
+      await handleInvalidPath("msg.error.backupPathContainsJoplinDir");
     }
```
[the remark this answers is not in the document]
Comment (personalizedrefrigerator) [reply src/Backup.ts]: I've changed this!

I'm not sure how necessary system folders are though — when I set the backup directory to `C:\ProgramFiles` or `C:\Windows`, I get an "operation not permitted" error.

[the remark this answers is not in the document]
Comment (JackGruber) [reply src/Backup.ts]: Yes when the user runs his system with UAC and or a administrator user.

**Review by JackGruber — APPROVED** [review]

**Pull request merged**